solvers/sffba.py

```diff
@@ -0,0 +1,284 @@
+            d_v=jnp.zeros_like(self.inner_var),
+            d_outer=jnp.zeros_like(self.outer_var),
+            state_lr= state_lr,
+            state_inner_sampler=self.state_inner_sampler,
+            state_outer_sampler=self.state_outer_sampler,
+        )
+        
+
+    def get_step(self, inner_sampler, outer_sampler):
+
+        grad_inner = jax.grad(self.f_inner, argnums=0)
+        grad_outer = jax.grad(self.f_outer, argnums=(0, 1))
+
+        def sffba_one_iter(carry, _):
+            r = 1
+            memory, carry = carry
+            (step_lr, inner_lr, outer_lr, assis_lr), carry['state_lr'] = update_lr(
+                carry['state_lr']
+            )
+            
+            # Step.1 - get all gradients and compute the implicit gradient.
+            (start_inner, id_inner, weight_inner,
+             carry['state_inner_sampler']) = (
+                inner_sampler(carry['state_inner_sampler'])
+            )
+
+            # previous gradient
+            grad_inner_var_old, vjp_train_old = jax.vjp(
+                lambda z, x: grad_inner(z, x, start_inner), carry['inner_var_old'],
+                carry['outer_var_old']
+            )
+            hvp_old, cross_v_old = vjp_train_old(carry['v_old'])
+
+            (start_outer, id_outer, weight_outer,
+             carry['state_outer_sampler']) = (
+                outer_sampler(carry['state_outer_sampler'])
+            )
+            grad_in_outer_old, grad_out_outer_old = grad_outer(
+                carry['inner_var_old'], carry['outer_var_old'], start_outer
+            )
+
+            carry['inner_var_old'] = carry['inner_var']
+            carry['outer_var_old'] = carry['outer_var']
+            carry['v_old'] = carry['v']
+
+
+            # this gradient
+            grad_inner_var, vjp_train = jax.vjp(
+                lambda z, x: grad_inner(z, x, start_inner), carry['inner_var'],
+                carry['outer_var']
+            )
+            hvp, cross_v = vjp_train(carry['v'])
+
+            grad_in_outer, grad_out_outer = grad_outer(
+                carry['inner_var'], carry['outer_var'], start_outer
+            )
+
+
+            # Step 2. - update the memory and performs the variance reduction
+            # here memory_*[-2] corresponds to the running average of
+            # the gradients and memory[-1] to the current direction
+            updates = {
+                'inner_grad': (grad_inner_var, id_inner, weight_inner),
+                'hvp': (hvp, id_inner, weight_inner),
+                'cross_v': (cross_v, id_inner, weight_inner),
+                'grad_in_outer': (grad_in_outer, id_outer, weight_outer),
+                'grad_out_outer': (grad_out_outer, id_outer, weight_outer),
+            }
+            memory = jax.tree_util.tree_map(
+                lambda mem, up: variance_reduction(mem, *up),
+                memory, updates
+            )
+            # Step.3 - update the direction
```

Comment:
If I got this right, this is an EMA? (not completely sure)
```suggestion
            # Step.3 - update the direction with an EMA
```

Reply:
A technique similar to EMA is employed here, but I believe simply changing it to "with EMA" may not fully capture the idea. In fact, it is a variant that incorporates memory variables. The specific update steps can be found in Section 3.2.1 of our paper. If you would like an explanation here, perhaps "with the stochastic estimator ZeroSARAH" would be more appropriate?